### PR TITLE
trafgen: l3: Support interface without IP address

### PR DESCRIPTION
**Pull request by abaw**
Hi,
I found the project this weekend and trafgen is exactly what I need at this moment. But playing with it, I found that it does not work with an interface without any IP address configured. For testing network equipments, I always remove the IP addresses of an interface to prevent Linux from sending unwanted packets(ARP, IGMP and so on) because those unwanted packets might affect my testing.

So I just fixed this by myself. Please take a look if you have time. Thanks for making this great tool.

### CODE REVIEW

**Comment by vkochan** [comment]
Right, I faced with this few days ago, but ... looks like a forgot about this. Thanks for point to this issue.

File: trafgen_l3.c [hunk 1]
```diff
@@ -81,6 +80,7 @@ static void ipv4_packet_finish(struct proto_hdr *hdr)
 
 	total_len = pkt->len - hdr->pkt_offset;
 	proto_field_set_default_be16(hdr, IP4_LEN, total_len);
+	proto_field_set_default_dev_ipv4(hdr, IP4_SADDR);
```
Comment (vkochan) [inline trafgen_l3.c]: I am thinking that may be it is better to handle failing of getting Ipv4/6 address from device and only warning about this
w/o panic.

File: trafgen_l3.c [hunk 2]
```diff
@@ -161,6 +160,7 @@ static void ipv6_packet_finish(struct proto_hdr *hdr)
 	uint16_t total_len = pkt->len - hdr->pkt_offset - IPV6_HDR_LEN;
 
 	proto_field_set_default_be16(hdr, IP6_LEN, total_len);
+  proto_field_set_default_dev_ipv6(hdr, IP6_SADDR);
```
Comment (vkochan) [inline trafgen_l3.c]: Looks like you forgot to use TAB here.

File: trafgen_l3.c (hunk 1 above)
Comment (vkochan) [inline trafgen_l3.c]: W/o digging into the code as far I remember by default the field value will be zeroed.

Comment (abaw) [inline trafgen_l3.c]: Good, I will check that tomorrow morning and  make necessary changes.

**Comment by tklauser** [comment]
Apart from the issues that Vadim addressed I think this is fine.

When you push your updated branch, could you please also include your Signed-off-by line in the commit as indicated in the [SubmittingPatches](https://github.com/netsniff-ng/netsniff-ng/blob/master/SubmittingPatches) document?

Thanks a lot for your contribution!

**Comment by abaw** [comment]
I made the suggested changes, please take a look. Thanks.

**Comment by tklauser** [comment]
LGTM. @vkochan, good to merge for you too?

File: trafgen_proto.c [hunk 3]
```diff
@@ -364,8 +364,10 @@ static void __proto_field_set_dev_ipv4(struct proto_hdr *hdr, uint32_t fid,
 		return;
 
 	ret = device_address(ctx.dev, AF_INET, &ss);
-	if (ret < 0)
-		panic("Could not get device IPv4 address\n");
+	if (ret < 0) {
+		fprintf(stderr,"Warning: Could not get device IPv4 address for %s\n", ctx.dev);
+		return;
+  }
```
Comment (vkochan) [inline trafgen_proto.c]: You need to shift '}' with a TAB to be aligned under 'if' on the same column.

File: trafgen_proto.c [hunk 4]
```diff
@@ -392,8 +394,10 @@ static void __proto_field_set_dev_ipv6(struct proto_hdr *hdr, uint32_t fid,
 		return;
 
 	ret = device_address(ctx.dev, AF_INET6, &ss);
-	if (ret < 0)
-		panic("Could not get device IPv6 address\n");
+	if (ret < 0) {
+		fprintf(stderr, "Warning: Could not get device IPv6 address for %s\n", ctx.dev);
+		return;
+  }
```
Comment (vkochan) [inline trafgen_proto.c]: Please see above's comment.

File: trafgen_proto.c (hunk 3 above)
Comment (tklauser) [inline trafgen_proto.c]: Good eye. Will fix when applying. Also the missing space after the comma...

File: trafgen_proto.c (hunk 4 above)
Comment (tklauser) [inline trafgen_proto.c]: Same here. Thanks @vkochan!

**Comment by abaw** [comment]
Thanks, I just fixed the tab and white space after the comma. And finally I learned to use checkpatch.pl from kernel source to check the coding style. Apart from these mentioned style problems, there are also lines exceeding 80 characters. Should I fix those long lines too?

**Comment by tklauser** [comment]
Picked and merged (with the spacing fixups) as c4e07d5142c8.

Thanks a lot @abaw and @vkochan!

**Pull request closed without merging**